File: tensorflow/python/data/experimental/kernel_tests/variant_test.py
# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""Tests for `tf.data.experimental.{from,to}_variant()`."""
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

from absl.testing import parameterized

from tensorflow.python.data.kernel_tests import test_base
from tensorflow.python.data.ops import dataset_ops
from tensorflow.python.framework import combinations
from tensorflow.python.platform import test


class VariantTest(test_base.DatasetTestBase, parameterized.TestCase):

  @combinations.generate(test_base.default_test_combinations())
  def testRoundtripRange(self):
    dataset = dataset_ops.Dataset.range(10)
    variant = dataset_ops.to_variant(dataset)
    dataset = dataset_ops.from_variant(variant,
                                       dataset_ops.get_structure(dataset))
    self.assertDatasetProduces(dataset, range(10))
    self.assertEqual(self.evaluate(dataset.cardinality()), 10)

  @combinations.generate(
      combinations.combine(tf_api_version=[2], mode=["eager", "graph"]))
  def testRoundtripMap(self):
    dataset = dataset_ops.Dataset.range(10).map(lambda x: x * x)
    variant = dataset_ops.to_variant(dataset)
    dataset = dataset_ops.from_variant(variant,
                                       dataset_ops.get_structure(dataset))
    self.assertDatasetProduces(dataset, [x * x for x in range(10)])
    self.assertEqual(self.evaluate(dataset.cardinality()), 10)


if __name__ == "__main__":
  test.main()

File: tensorflow/lite/core/async/interop/variant_test.cc
/* Copyright 2022 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/
#include "tensorflow/lite/core/async/interop/variant.h"

#include <cstddef>
#include <string>

#include <gtest/gtest.h>

namespace tflite::interop {
namespace {

TEST(VariantTest, IntTest) {
  {
    Variant a(1);
    EXPECT_EQ(1, *a.Get<int>());
  }
  {
    Variant a(1);
    a.Set(2);
    EXPECT_EQ(2, *a.Get<int>());
  }
  {
    Variant a(42);
    Variant b(a);
    EXPECT_EQ(42, *b.Get<int>());
  }
  {
    Variant a(42);
    EXPECT_EQ(42, *static_cast<const int*>(a.GetPtr()));
  }
  {
    Variant a(42);
    Variant b(42);
    EXPECT_EQ(a, b);
    b.Set(21);
    EXPECT_NE(a, b);
  }
}

TEST(VariantTest, SizeTTest) {
  {
    size_t v = 1;
    Variant a(v);
    EXPECT_EQ(1, *a.Get<size_t>());
  }
  {
    size_t v = 1;
    Variant a(v);
    size_t t = 2;
    a.Set(t);
    EXPECT_EQ(2, *a.Get<size_t>());
  }
  {
    size_t v = 42;
    Variant a(v);
    Variant b(a);
    EXPECT_EQ(42, *b.Get<size_t>());
  }
  {
    size_t v = 42;
    Variant a(v);
    EXPECT_EQ(42, *static_cast<const size_t*>(a.GetPtr()));
  }
  {
    Variant a(size_t(42));
    Variant b(size_t(42));
    EXPECT_EQ(a, b);
    b.Set(size_t(21));
    EXPECT_NE(a, b);
  }
}

TEST(VariantTest, StringTest) {
  {
    const char v[] = "string";
    Variant a(v);
    EXPECT_EQ(v, *a.Get<const char*>());
  }
  {
    const char v[] = "string";
    Variant a(v);
    const char t[] = "another string";
    a.Set(t);
    EXPECT_EQ(t, *a.Get<const char*>());
  }
  {
    const char v[] = "string";
    Variant a(v);
    Variant b(a);
    EXPECT_EQ(v, *b.Get<const char*>());
  }
  {
    const char v[] = "string";
    Variant a(v);
    EXPECT_EQ(v, *static_cast<const char* const*>(a.GetPtr()));
  }
  {
    const char v[] = "string";
    Variant a(v);
    std::string str = "string";
    Variant b(str.c_str());
    EXPECT_EQ(a, b);
    b.Set("another string");
    EXPECT_NE(a, b);
  }
}

TEST(VariantTest, TypeNotMatch) {
  Variant a(1);
  EXPECT_EQ(nullptr, a.Get<size_t>());
}

}  // namespace
}  // namespace tflite::interop
